---
title: "data_setup"
author: "Micah GVI"
date: "04/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r package load, warning=FALSE, message=FALSE}
library(BiocManager)
library(annotate)
library(ape)
library(ggtree)
library(seqinr)
library(Biostrings)
library(wesanderson)
library(viridis)
library(ggplot2)
library(stringi)
library(treeio)
library(reshape2)
library(phytools)
library(mapdata)
```

### Amino acid sequence code ####
```{r data import, echo=FALSE}
seq <- readAAStringSet("./input/master_aa_seqs.fa")
seqDF <- data.frame(seq)
names(seqDF) <- c("seq")
```


```{r}
#Extract sequence titles to format/add information
title = rownames(seqDF)
title = as.data.frame(title)

write.csv(names, "./input/title.csv", row.names=FALSE)

#Missing Vector and Country information is added in here using accession literature 
#File is re-saved as titlefinal

WNV_Variant = read.csv("./input/titlefinal.csv")

seqDF = cbind("WNV Variant" = WNV_Variant, seqDF)
rownames(seqDF) = seqDF[,1]
```


```{r}
## Data cleanup/removal of abberant seqs
# Check sequence lengths.
SeqLen<-as.numeric(lapply(seqDF$seq,nchar))
qplot(SeqLen, binwidth = 50)+theme_bw() # There are nine sequences that are drastically different lengths than the rest. I will remove these:
KeepSeq<-SeqLen>3000
seqDF<-seqDF[KeepSeq,]


## Isolating structural proteins.

cpro_seq<- unlist(substr(seqDF$seq, start = 6, stop = 123))
cproAA<-sapply(cpro_seq,strsplit,split="")
names(cproAA) <- seqDF$Title
cproAA <- as.AAbin(cproAA)


prMpro_seq <- unlist(substr(seqDF$seq, start = 217, stop = 290))
prMproAA<-sapply(prMpro_seq,strsplit,split="")
names(prMproAA) <- seqDF$Title
prMproAA <- as.AAbin(prMproAA)


epro_seq <- unlist(substr(seqDF$seq, start = 598, stop = 791))
eproAA<-sapply(epro_seq,strsplit,split="")
names(eproAA) <- seqDF$Title
eproAA <- as.AAbin(eproAA)
```

### Function for Alignment and Distance Matrices
```{r DM function} 
DMfunction <- function(protein) {
  align<-muscle(protein,quiet=F)
  
  alignmentvisual <- image.AAbin(align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)
  
  DM<-dist.aa(align) # generate distance matrix based on AA sequences.
  DMmat<-as.matrix(DM) # convert to matrix format.
dim(DMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
DMlmat<-melt(DMmat) # Linear matrix.
ggplot(data=DMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("C Protein Distance Matrix")
}

DMfunction(cproAA)
DMfunction(prMproAA)
DMfunction(eproAA)
```


```{r protein sequence alignments}
## C protein muscle aligment ## 
cproAA_align<-muscle(cproAA,quiet=F)
cpro_alignment_visual <- image.AAbin(cproAA_align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)

## prM protein muscle alignment ##
prMproAA_align<-muscle(prMproAA,quiet=F)
prMpro_alignment_visual <- image.AAbin(prMproAA_align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)

## E protein muscle alignment ## 
eproAA_align<-muscle(eproAA,quiet=F)
epro_alignment_visual <- image.AAbin(eproAA_align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)

```

```{r distance matrices}
## C protein distance matrix. ##
CproDM<-dist.aa(eproAA_align) # generate distance matrix based on AA sequences.
CproDMmat<-as.matrix(CproDM) # convert to matrix format.
dim(CproDMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
CproDMlmat<-melt(CproDMmat) # Linear matrix.
CproDMvis<-ggplot(data=CproDMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("C Protein Distance Matrix")

## prM protein distance matrix. ##
prMproDM<-dist.aa(prMproAA_align) # generate distance matrix based on AA sequences.
prMproDMmat<-as.matrix(prMproDM) # convert to matrix format.
dim(prMproDMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
prMproDMlmat<-melt(prMproDMmat) # Linear matrix.
prMproDMvis<-ggplot(data=prMproDMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("prM Protein Distance Matrix")


## E protein distance matrix. ##
eproDM<-dist.aa(eproAA_align) # generate distance matrix based on AA sequences.
eproDMmat<-as.matrix(eproDM) # convert to matrix format.
dim(eproDMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
eproDMlmat<-melt(eproDMmat) # Linear matrix.
eproDMvis<-ggplot(data=eproDMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("E Protein Distance Matrix")

```

### Old code ###
```{r initial alignments}
#seqAlign<-muscle(seqAA,quiet=F)
#alview(seqAlign)
#monopic <- image.AAbin(seqAlign, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)
```

## Group Vector Information and create Phylogeny

### Create a function to make a phylogenetic tree for proteins 
```{r}
# this functions takes in the protein for the tree to be created and whether it should create a phylogenetic tree for country or vector 

treefunction <- function(tree, type) {
  PhyloTree<- nj(tree)
  
  country = stri_extract_all(PhyloTree$tip.label, regex= "USA|Australia|South Africa|Israel|Russia|India|Madagascar|Azerbaijan|Ukraine|Czech Republic|Kenya|Nigeria|Central African Republic|DRC|Portugal|Morocco|Senegal|Tunisia|Cyprus|Japan|France", mode = "all", simplify =TRUE)  

  vect = stri_extract_all(PhyloTree$tip.label, regex= "Culex|Other|Equus|Homo sapiens|Corvus|Culiseta|Ornithodoros|Mus|Hyalomma|Oriolus|Aves|Dermacentor|Aedes|Rousettus|Culicidae",mode = "all", simplify =TRUE) 

if (type == "country"){
Groups = split(PhyloTree$tip.label, country) #Separate country from strain name
}
else if (type == "vect"){
Groups = split(PhyloTree$tip.label, vect) #Separate vector from strain name
}
col = groupOTU(PhyloTree,Groups) #Creates a group
ggtree(col, layout="rectangular", branch.length = "none", aes(color = group))#creates tree  
}

```

```{r}
#c protein trees
treefunction(CproDM, "country")
treefunction(CproDM, "vect")

#prM protein trees
treefunction(prMproDM, "country")
treefunction(prMproDM, "vect")

#e protein trees 
treefunction(eproDM, "country")
treefunction(eproDM, "vect")
```

### Zooming in on interesting parts of each phylogeny

```{r}
#Create function to generate new tip labels for each tree
nametip = function(tree) {
  
#Use regex to select country out of the Tip labels
country = stri_extract_all(tree$tip.label, regex= "USA|Australia|South Africa|Israel|Russia|India|Madagascar|Azerbaijan|Ukraine|Czech Republic|Kenya|Nigeria|Central African Republic|DRC|Portugal|Morocco|Senegal|Tunisia|Cyprus|Japan|France", mode = "all", simplify =TRUE) 

#Use regex to select years out of the E Protein Tip labels
Year = stri_extract_all(tree$tip.label, regex = "\\d{4}[^\\d]*$",mode = "all", simplify =TRUE)

#Combine the years and countries together and set those values as the new tip labels (Country,Year)
name = melt(data.frame(country, Year))
name$full = paste(name$country, name$Year, sep=",")

return(name$full)
}
```

```{r}
TreeE = nj(eproDM) #Create Tree using E Protein
TreeC = nj(CproDM) #Create Tree using E Protein
TreeprM = nj(prMproDM) #Create Tree using prM protein

TreeE$tip.label = nametip(TreeE) #Assign new tip labels to each protein tree
TreeC$tip.label = nametip(TreeC)
TreeprM$tip.label = nametip(TreeprM)
```

```{r}
#Use ggtree to generate tree from E Protein
ET = ggtree(TreeE, branch.length = "none")+
  geom_tiplab(size = 2.75)+ #Decrease tip label size
  xlim(NA,60)+  #Shift phylogeny to the left
  geom_hilight(node=164, fill="gold", alpha = 0.15) + #highlight distinct groupings in phylogeny
  geom_hilight(node=145, fill="purple", alpha = 0.15)+ 
  geom_hilight(node=135, fill="blue", alpha = 0.15)

ET = ggtree::collapse(ET, node=153) #collapse clade to the right of groupings

viewClade(ET, MRCA(ET, 82, 80)) #Zoom in on interesting groupings
```

```{r}
#Use ggtree to generate tree from C Protein
CT = ggtree(TreeC, branch.length = "none")+
  geom_tiplab(size = 2.75)+ #Decrease tip label size
  xlim(NA,60)+  #Shift phylogeny to the left
  geom_hilight(node=164, fill="orange", alpha = 0.2) + #highlight distinct groupings in phylogeny
  geom_hilight(node=145, fill="pink", alpha = 0.2)+ 
  geom_hilight(node=135, fill="skyblue", alpha = 0.2)

CT = ggtree::collapse(CT, node=153) #collapse clade to the right of groupings

viewClade(CT, MRCA(CT, 82, 80)) #Zoom in on interesting groupings
```

```{r}
#Use ggtree to generate tree from prM Protein
prMT = ggtree(TreeprM, branch.length = "none")+
  geom_tiplab(size = 2.75)+ #Decrease tip label size
  xlim(NA,80)+ #Shift phylogeny to the left
  geom_hilight(node=144, fill="gold", alpha = 0.15) + #highlight distinct groupings in phylogeny
  geom_hilight(node=161, fill="purple", alpha = 0.15)+ 
  geom_hilight(node=174, fill="blue", alpha = 0.15)

viewClade(prMT, MRCA(prMT, 140, 80)) #Zoom in on interesting groupings
```








