---
title: "tree code"
output: html_notebook
---

```{r}
str(seqDM)
NJtree<-nj(seqDM)
ggtree(NJtree, layout="rectangular")

MEtree<-fastme.bal(seqDM)
ggtree(MEtree)

head(PhyloTree$tip.label)
```

**I am not too sure what stuff is redundant here so I will leave it to someone else to clean up!)**

```{r NJ phylogenetic tree setup}
PhyloTree = nj(seqDM) #Build a phylogeny using neighbour-joining method <- <- 
```

```{r group by country}

#Use regex and stringi to select strain vector from dataset
Vect = stri_extract_all(PhyloTree$tip.label, regex= "Culex|Homo sapiens|Culiseta|Corvus|Aves|Aedes|Culicidae",
                        mode = "all", simplify =TRUE) 
#Vect<-sub("\\w+\\s+\\|(\\w+|\\w+\\s\\w+)\\|(\\w+|\\w+\\s\\w+)\\|.+","\\1",PhyloTree$tip.label) #uses regex to capture all vectors. Only keeping here if needed in the future analysis

Groups = split(PhyloTree$tip.label, Vect) #Separate vector from strain name

col = groupOTU(PhyloTree,Groups) #Create a group composed of vectors

str(col) #Display "col" information to ensure groups were properly formed

```


```{r country tree}
ggtree(col, layout="rectangular", aes(color = group))

#Two genomes were missing the country, so I checked the research papers where the data was published and added in the missing countries to our dataset (Kenya and Czech Republic)

Country = stri_extract_all(PhyloTree$tip.label, regex= "USA|Australia|South Africa|Israel|Russia|India|Madagascar|Azerbaijan|Ukraine|Czech Republic|Kenya", mode = "all", simplify =TRUE) 

#Country<-sub("\\w+\\s+\\|(\\w+|\\w+\\s\\w+)\\|(\\w+|\\w+\\s\\w+)\\|.+","\\2",PhyloTree$tip.label) #uses regex to capture country 

CountryGroups = split(PhyloTree$tip.label, Country)
Ccol = groupOTU(PhyloTree,CountryGroups)
str(Ccol) #Extra group category 0 is being created for some reason

ggtree(Ccol, layout="rectangular", branch.length = "none", aes(color = group))
```


### Minimum Evolution Tree
```{r ME Country Tree}
MEtree<-fastme.bal(seqDM)

Country1<-sub("\\w+\\s+\\|(\\w+|\\w+\\s\\w+)\\|(\\w+|\\w+\\s\\w+)\\|.+","\\2",MEtree$tip.label) #uses regex to capture country 

CountryGroups1 = split(MEtree$tip.label, Country)
MECcol = groupOTU(MEtree,CountryGroups1)
str(MECcol)
ggtree(MECcol, layout="rectangular",branch.length = "none", aes(color = group))
```

##Create a function to make a phylogenetic tree for proteins 
```{r}
treefunction <- function(tree, type) {
  PhyloTree<- nj(tree)
  country = stri_extract_all(PhyloTree$tip.label, regex= "USA|Australia|South Africa|Israel|Russia|India|Madagascar|Azerbaijan|Ukraine|Czech Republic|Kenya", mode = "all", simplify =TRUE)
  vect = stri_extract_all(PhyloTree$tip.label, regex= "Accipiter|Corvus|Aves|Accipitridae|Anser|Aquila|Branta|Bubo|Buteo|Butorides|Cardinalis|Carpodacus|Ciconiidae|Colaptes|Columbia|Columbidae|Corvidae|Cyanocitta|Dumetella|Falco|Gallus|Laridae|Mimus|Passer|Passeridae|Passeriformes|Phalacrocorax|Phoenicopterus|Pica|Poecile|Quiscalus|Sciuridae|Turdus|Zenaida|Columba",mode = "all", simplify =TRUE) 

if (type == "country"){
Groups = split(PhyloTree$tip.label, country) #Separate country from strain name
}
else if (type == "vect"){
Groups = split(PhyloTree$tip.label, vect) #Separate vector from strain name
}
col = groupOTU(PhyloTree,Groups) #Create a group composed of vectors
str(col) #Display "col" information to ensure groups were properly formed
ggtree(col, layout="rectangular", aes(color = group))
}

```

```{r}
#c protein trees
treefunction(CproDM, country)
treefunction(CproDM, vect)

#prM protein trees
treefunction(prMproDM, country)
treefunction(prMproDM, vect)

#e protein trees 
treefunction(eproDM, country)
treefunction(eproDM, vect)
```


```{r protein sequence alignments}
## C protein muscle aligment ## 
cproAA_align<-muscle(cproAA,quiet=F)
cpro_alignment_visual <- image.AAbin(cproAA_align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)

## prM protein muscle alignment ##
prMproAA_align<-muscle(prMproAA,quiet=F)
prMpro_alignment_visual <- image.AAbin(prMproAA_align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)

## E protein muscle alignment ## 
eproAA_align<-muscle(eproAA,quiet=F)
epro_alignment_visual <- image.AAbin(eproAA_align, bg = (wes_palette("Darjeeling1")[5]), col = wes_palette("Darjeeling1"), ylab = "WNV Variants", xlab = "Sequence Length", show.labels = T, cex.lab = 0.5, legend = T, aa.cex = 0.8)

```

```{r distance matrices}
## C protein distance matrix. ##
CproDM<-dist.aa(proAA_align) # generate distance matrix based on AA sequences.
CproDMmat<-as.matrix(CproDM) # convert to matrix format.
dim(CproDMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
CproDMlmat<-melt(CproDMmat) # Linear matrix.
CproDMvis<-ggplot(data=CproDMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("C Protein Distance Matrix")

## prM protein distance matrix. ##
prMproDM<-dist.aa(prMproAA_align) # generate distance matrix based on AA sequences.
prMproDMmat<-as.matrix(prMproDM) # convert to matrix format.
dim(prMproDMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
prMproDMlmat<-melt(prMproDMmat) # Linear matrix.
prMproDMvis<-ggplot(data=prMproDMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("prM Protein Distance Matrix")


## E protein distance matrix. ##
eproDM<-dist.aa(eproAA_align) # generate distance matrix based on AA sequences.
eproDMmat<-as.matrix(eproDM) # convert to matrix format.
dim(eproDMmat) # 104 x 104

# Melt to a linear matrix and generate distance matrix visual.
eproDMlmat<-melt(eproDMmat) # Linear matrix.
eproDMvis<-ggplot(data=eproDMlmat,aes(x=Var1,y=Var2,fill=value)) + geom_tile() + scale_fill_gradientn(colours=c("white","blue","green","red")) + theme(text = element_text(size=5),
        axis.text.x = element_text(angle=90, hjust=1)) + ggtitle("E Protein Distance Matrix")

```
