---
title: "ORF Grabbing"
output: html_notebook
---

Big success in grabbing the ORFs
```{r}
sample_seq<-read.fasta("./input/sample_sequence.fasta")
sample_seq <- unlist(sample_seq, use.names = F)
sample_seq <- paste(sample_seq, collapse = "")
sample_seq<-gsub("([actg])","\\U\\1",sample_seq,perl=T)

m <- gregexpr("(ATG([ACTG]{3,3})+?((TAA)|(TAG)|(TGA)))",sample_seq)
hi <- regmatches(sample_seq,m)
hi
```

Unfortunately this reveals that WNV is made of a polyprotein (i.e. a single massive ORF) that is cleaved in order to produce the individual proteins. This means we will have to approximate the Amino Acid position ranges for each protein. To do this we will use Amino Acid sequences, and isolate the regions noted to be part of "conserved domains" that are specific to each protein as dictated by NCBI's Conserved Domain Database...

```{r}
seq <- readAAStringSet("./input/master_aa_seqs.fa")
seqDF <- data.frame(seq)
names(seqDF) <- c("seq")
seqDF <- cbind("WNV Variant" = rownames(seqDF), seqDF)

cpro_seq<- unlist(substr(seqDF$seq, start = 6, stop = 123))
cprodf <- cbind("WNV Variant" = rownames(seqDF), cpro_seq)
cprodf <- as.data.frame(cprodf)
cproAA<-sapply(cprodf$cpro_seq,strsplit,split="")
names(cproAA) <- seqDF$`WNV Variant`
cproAA <- as.AAbin(cproAA)

prMpro_seq <- unlist(substr(seqDF$seq, start = 217, stop = 290))
prMprodf <- cbind("WNV Variant" = rownames(seqDF), prMpro_seq)
prMprodf <- as.data.frame(prMprodf)
prMproAA<-sapply(prMprodf$prMpro_seq,strsplit,split="")
names(prMproAA) <- seqDF$`WNV Variant`
prMprodf_AA <- as.AAbin(prMproAA)

epro_seq <- unlist(substr(seqDF$seq, start = 598, stop = 791))
eprodf <- cbind("WNV Variant" = rownames(seqDF), epro_seq)
eprodf <- as.data.frame(eprodf)
eproAA<-sapply(eprodf$epro_seq,strsplit,split="")
names(eproAA) <- seqDF$`WNV Variant`
eprodf_AA <- as.AAbin(eproAA)


```

